### PR TITLE
[2pt] PR: Additional new model source code to system

### LINKED ISSUE #254: [8pt] Add source code to ras2fim tools
During some design reviewing it was discovered that we have a need to know and use a code to identify the source.  At this point, we have only one source provider which is BLE. We will use the `source_code` of `ble` for the first records. The code is needed for a number of reasons including merging when multiple runs of ras2fm.py for one HUC are created and also HydroVIS wants this value as well. 

The unit file name will add the code in the folder name. What was {huc8}_{crs number}_{date}, ie) 12090301_2276_240126, will now become {huc8}_{crs number}_{source code}_date, ie) 12090301_2276_ble_240126

Some minor modifications are required for this fix:
1) ras2fim.py  (for creating the folder name)
2) get_models_by_catalog.py (for adding source code as an argument to find the right model records for the source.
3) ras_unit_to_s3.py  (to search for duplication s3 output folder to include the new source code value.

This new file system will be used in the pending tools for building a huc release folder (unit_builder) as well as ras2release.py.

### PR DESCRIPTION
**Pull request by RobHanna-NOAA**
During some design reviewing it was discovered that we have a need to know and use a code to identify the source.  At this point, we have only one source provider which is BLE. We will use the `source_code` of `ble` for the first records. The code is needed for a number of reasons including helping with merging when multiple runs of ras2fm.py for one HUC are created and also HydroVIS wants this value as well. 

The unit file name will add the code in the folder name. What was `{huc8}_{crs number}_{date}`, ie) 12090301_2276_240126, will now become `{huc8}_{crs number}_{source code}_{date}`, ie) 12090301_2276_ble_240126.

This triggered the addition of a new system to know a source code value such as "ble" and get a human friendly value which HydroVIS also wants. ie) ble = FEMA Base Level Engineering (BLE). This has to be done at this level as HV won't know the source so can not translate it as other sources come online. These value will be added to the unit output rating curves for HV.

A number of files are also be removed from the tools folder that are no longer used.

**NOTE: `tools\get_models_by_catalog.py` has been renamed here to `tools\s3_get_models.py`**

Closes Issue [254](https://github.com/NOAA-OWP/ras2fim/issues/254)

### Additions  

- `config`
    - `source_codes.csv`

### Renamed Files
- `tools`
    -  Was `get_models_by_catalog.py`, now `s3_get_models.py`

### Changes  

- `pyproject.toml`: minor changes due to file renamed.
- `src`
    - `ras2fim.py`: Addition of code in support of the new `source_code` parameter.
    - `ras2fim_logger.py`: Fix bug were it would creating a full path instead of just the last folder.
    - `shared_functions.py`: Addition of code in support of the new `source_code` parameter.
    - `shared_variables.py`: Renamed some variables related to S3 objects.
- `tools`
    - `ras_unit_to_s3.py`: Addition of code in support of the new `source_code` parameter. Some variable names and text adjustments.
    - `s3_get_models.py`: Addition of code in support of the new `source_code` parameter. Some variable names and text adjustments.
    - `s3_model_mgmt.py`: Text changes.
    - `s3_search_tool.py`: Text changes.
    - `s3_shared_functions.py`: Addition of code in support of the new `source_code` parameter.

### Removals 

- `tools`
    - `convert_ras2fim_to_recurr_validation_datasets.py`
    - All files starting with `nws_`

### Testing

- **_For review testing,_** I recommend the following steps. There are a ton of possible tests. See list below of the ones I did. If you are comfortable, you can do a shortened version.
     - Pulling down models for any huc (but not 1209301)
     - Make a smaller subset model folder of that huc.
     - Run it through ras2fim.py. It doesn't matter that it doesn't finish (v2 not ready yet).
     - Use `ras_unit_to_s3.py` to upload it, _but make sure you set the -b arg as_ `ras2fim-dev`. Ensure it actually did upload to the `output_ras2fim` directory. Download the `output_ras2fim\ras_output_tracker.csv` and ensure it has a new entry showing the upload. You are welcome to more testing. Below I will list all tests that I performed.

- **_Testing done during development_**
   
Start with a new download of data using s3_get_models.py.
	 - Make a temp models folders with a small subset of the models you just download. These test are not about sizes but folder names, so if you keep your  folder sizes small, it will much quicker to test. As long as it has a "final" folder, it will allow you to upload.
	
	- Run ras2fim against it, the ras2fim from this code base which will add the source code to the output folder name. It will fail as expected after Step 6, reset is still WIP. Add a subfolder named "final" if it doesn't already exist.

Now we can test ras_unit_to_s3.py features.
Note: Ensure all bucket input args in ras_unit_to_s3.py values are set to "ras2fim-dev"  (using the -b arg)

	- Ensuring that the combination of huc, crs and source code does not exists in s3://ras2fim-dev/output_ras2fim/ (just delete it in S3 ras2fim-dev if one it does exist. At this point, the date in the folder is irrelevant.
	
	- Note: For each time a unit folder is uploaded or moved into S3, it will update the tracker file at s3://ras2fim-dev/output_ras2fim/ras_output_tracker.csv. Each time I do another ras_unit_to_s3.py, I download that file to local and see if it added records of the transactions correctly. That file has some old historical data it in, so just worry about new records being added as steps are run below.

	- run ras_unit_to_s3.py to push up your new ras output folder.  It should load straight up with no  questions.
	
Next we are going to test duplicate folders in S3 with the exact same unit folder name, including date in folder name.
	- Open your unit folder and add some sort of temp file in the root unit folder. Doesn't matter what the file is (big or small). We will watch for that file shortly.
	
	- run ras_unit_to_s3.py with the exact args as above. Type "abort" when prompted.
	
	- run ras_unit_to_s3.py with the exact args as above. It should see you have an exact duplicate of the unit folder. Type "overwrite" when prompted.
	
	- Open your unit folder and add some another of temp file in the root unit folder. Doesn't matter what the file is (big or small). We will watch for that file shortly.
	
	- run ras_unit_to_s3.py with the exact args as above again. It should see you have an exact duplicate of the unit folder. This time, type "archive". You should see the existing folder moved into the "output_ras2fim_archive" folder. It will add a date/time to the folder name. You should not see your temp file in the archived version but should see it the output_ras2fim folder version. This helps to ensure the right unit folder was saved in the right place.

Using your original local unit folder, rename it by changing it date to a newer date. Now we can test newer versions of a unit coming in from what S3 already 
has. 
	- Ensure your script has the new folder name.
	
	- Rename our previously created temp file to anything. Remove the temp files you created earlier.
	
	- run ras_unit_to_s3.py with the newer dated local unit folder name.  Type "abort" when prompted.
	
	- run ras_unit_to_s3.py  with the exact args as above. Type "incoming" and watching that temp(s) file and date/times in the archive folder, you should see the new incoming one go straight to the archive folder.
	
	- check the archive folder new version to see it has the renamed temp file. It helps validate that the new incoming version did go straight to archive.
	
	- Add a new temp file to our source unit folder.
	
	- run ras_unit_to_s3.py with the exact same params, but this time type "existing". This time, you should see the one that exists in output_ras2fim (with the older date), gets moved to archives and the new incoming gets uploaded to outputs_ras2fim. Keeping an eye on the temp file name will help validate which folder was moved and which was uploaded.

	- Upload a new huc folder into output_ras2fim. Prior to the run, ensure there is no already existing folder with the same huc/crs/souce code. Should load up with no issues and no questions.

### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [ ] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [ ] Pre-commit executed and linting changes made.
- [ ] Links are provided if this PR resolves an issue, or depends on another other PR
- [ ] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [ ] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated ([CHANGELOG](/doc/CHANGELOG.md) and/or [README](/README.md))
- [ ] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)

### Merge Checklist (For Technical Lead use only)

- [ ] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number and merge date

### CODE REVIEW

File: src/ras2fim.py [hunk 1]
```diff
@@ -552,8 +562,8 @@ def create_input_args_log(**kwargs):
     #            -n E:\X-NWS\X-National_Datasets -r "C:\Program Files (x86)\HEC\HEC-RAS\6.3"
```
Comment (JamesColl-NOAA) [inline src/ras2fim.py:569]: I believe this line needs to be tweaked slightly to include the new required inputs

**Review by JamesColl-NOAA — DISMISSED** [review]
Lightly tested compared to your docs but this added source to outputs and looks like it's merged correctly.  Approved and we'll continue to find edge cases to test!

**Review by CarsonPruitt-NOAA — APPROVED** [review]
